tokio-util/src/sync/mpsc.rs

```diff
@@ -64,13 +64,50 @@ async fn make_acquire_future<T>(
     }
 }
 
-impl<T: Send + 'static> PollSender<T> {
+type InnerFuture<'a, T> = ReusableBoxFuture<'a, Result<OwnedPermit<T>, PollSendError<T>>>;
+
+#[derive(Debug)]
+// TODO: This should be replace with a type_alias_impl_trait to eliminate `'static` and all the transmutes
+struct PollSenderFuture<T>(InnerFuture<'static, T>);
+
+impl<T> PollSenderFuture<T> {
+    /// Create with an empty inner future with no `Send` bound.
+    fn empty() -> Self {
+        // We don't use `make_acquire_future` here because our relaxed bounds on `T` are not
+        // compatible with the transitive bounds required by `Sender<T>`.
+        Self(ReusableBoxFuture::new(async { unreachable!() }))
+    }
```

Comment:
Can you post the error you get when trying to do that?

The reason I used a `make_acquire_future(None)` call was because this results in a future of the same size and alignment as the actual futures, meaning that the allocation made here would be reused.

Perhaps a solution is to drop the `Send` requirement too?

Reply:
```
error[E0277]: `T` cannot be sent between threads safely
  --> tokio-util/src/sync/mpsc.rs:78:37
   |
78 |         Self(ReusableBoxFuture::new(make_acquire_future(None)))
   |              ---------------------- ^^^^^^^^^^^^^^^^^^^^^^^^^ `T` cannot be sent between threads safely
   |              |
   |              required by a bound introduced by this call
```

I feel less good about casting away `Send` on a `Box`, since that's not just a lifetime cast.

This isn't actually a change, just a refactor, as this is still only used in `PollSender::clone`. I just moved it to consolidate the comments on the `ReusableBoxFuture` usage to be internal to `PollSenderFuture`.

In the future `type_alias_impl_trait` implementation, these two functions can be the same as the option can be outside the future instead of inside, and there will be no `Box`, so no need to consider reallocation.